### PR TITLE
Publication updater

### PR DESCRIPTION
**Pull request by daisieh**
Installs a new webapp that checks workflow items and updates them with manuscript numbers from journal metadata and publication metadata from CrossRef. Updated items are summarized and sent as an email to the specified curator account.

### CODE REVIEW

**Comment by daisieh** [comment]
To install this PR:
- Import some version of the database with lots of entries in it (`import_pg_dump.sh` on the VM)
- Update dryad schema: `/opt/dryad/bin/dsrun org.dspace.administer.MetadataImporter -f /opt/dryad/config/registries/dryad-types.xml`
- Set the email account in the maven settings: `<default.curator.recipient>dryadassistant@datadryad.org</default.curator.recipient>`

To test:
- Redeploy Dryad.
- Run `run_mailserver.sh` on the VM.
- Observe emails being received for each journal:
- Items listed with "Associated publication (match score) was found" should have new metadata and provenance added: `dc.relation.isreferencedby`, `dryad.citationInProgress`, `dc.date.issued`. There should also be a filled-in `dc.identifier.citation` with approximately correct formatting (curators will manually verify this). Any associated journal metadata entry in the manuscript database will also be updated with the publication_doi and status "published".
- Items listed with "Journal-provided metadata for msid xxxx-xxxx with title 'blah blah' was added." should have whatever msid updated in `dc.identifier.manuscriptNumber`.

File: dspace/modules/api/src/main/java/org/datadryad/rest/storage/rdbms/ManuscriptDatabaseStorageImpl.java [hunk 1]
```diff
@@ -157,25 +163,81 @@ static TableRow tableRowFromManuscript(Manuscript manuscript, Integer organizati
...
+            if (!"".equals(authorString.toString())) {
+                String query = "SELECT * FROM MANUSCRIPT WHERE organization_id = ? and active = ? " + authorString.toString();
+                TableRowIterator tableRowIterator = DatabaseManager.queryTable(context, MANUSCRIPT_TABLE, query, organizationId, ACTIVE_TRUE);
+                if (tableRowIterator != null) {
+                    List<TableRow> rows = tableRowIterator.toList();
+                    for (TableRow row : rows) {
+                        Manuscript databaseManuscript = manuscriptFromTableRow(row);
+                        String databaseTitle = StringUtils.deleteWhitespace(StringUtils.upperCase(databaseManuscript.getTitle()));
+                        String manuscriptTitle = StringUtils.deleteWhitespace(StringUtils.upperCase(manuscript.getTitle()));
+                        double score = JournalUtils.getHamrScore(databaseTitle, manuscriptTitle);
```
Comment (ryscher) [inline dspace/modules/api/src/main/java/org/datadryad/rest/storage/rdbms/ManuscriptDatabaseStorageImpl.java:227]: Oy! This is the first time HAMR will be used for real production activities!

**Comment by ryscher** [comment]
For production deploy, need to set maven param for default.publication.updater.timer

**Comment by ryscher** [comment]
Also need to add an entry to tomcat server.xml to deploy the webapp.

**Comment by ryscher** [comment]
A package called "Data from: test" matched something with a much longer title, like "Test tube blah blah and some such". This may not be avoidable, but it seems a little aggressive.

**Comment by daisieh** [comment]
Yeah, overly short things that are test packages still stuck in workflow are probably going to get caught, but since there's no real harm in writing metadata to those test package, this shouldn't be a problem.

**Comment by ryscher** [comment]
Run with e.g., http://mydryadserver.com/publication-updater/retrieve

**Comment by daisieh** [comment]
I don't understand...that already works here.

File: dspace/modules/publication-updater/src/main/java/org/datadryad/publication/PublicationUpdater.java [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+    private final static String PROVENANCE = "dc.description.provenance";
+
+    // Timer for scheduled harvesting of publications from crossref
+    private Timer myPublicationUpdaterTimer;
+
+    @Override
+    protected void doGet(HttpServletRequest aRequest,
+                         HttpServletResponse aResponse) throws ServletException, IOException {
+        String requestURI = aRequest.getRequestURI();
+        if (requestURI.contains("retrieve")) {
```
Comment (daisieh) [inline dspace/modules/publication-updater/src/main/java/org/datadryad/publication/PublicationUpdater.java:58]: retrieve is here

**Comment by ryscher** [comment]
In all of the items I've checked so far, the provenance field is duplicated.

**Comment by ryscher** [comment]
The dc.relation.isreferencedby needs a prefix of "doi:"

**Comment by daisieh** [comment]
Are they duplicated exactly? Or is there one for "Journal-provided metadata was updated" and then "Associated publication" with the same timestamps?

**Comment by ryscher** [comment]
Needs lots of documentation cleanup, but the code looks good to go!

**Pull request merged**